test/test_arrays.jl

```diff
@@ -98,56 +240,102 @@ end
     pow3(A) = A^3
 
     # Functions that only work on square matrices
-    @testset "$name" for (name, A) in TEST_SQUARE_MATRICES
-        test_patterns(inv, A; outsum=true)
-        test_patterns(exp, A; outsum=true)
-        test_patterns(pow0, A; outsum=true, con=iszero, jac=iszero, hes=iszero)
-        test_patterns(pow3, A; outsum=true)
+    @testset "inv $(arrayname(A))" for A in NONDIAG_SQUARE_MATRICES
+        testJ1(inv, A)
+        testH1(inv, A)
+    end
+    @testset "inv $(arrayname(A))" for A in DIAG_SQUARE_MATRICES
+        @test Hsum(inv, A) == [
+            1  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  1  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  1
+        ]
+    end
+    @testset "exp $(arrayname(A))" for A in NONDIAG_SQUARE_MATRICES
+        testJ1(exp, A)
+        testH1(exp, A)
+    end
+    @testset "exp $(arrayname(A))" for A in DIAG_SQUARE_MATRICES
+        testJ1(exp, A)
+        @test Hsum(exp, A) == [
+            1  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  1  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  0
+            0  0  0  0  0  0  0  0  1
+        ]
+    end
+    @testset "pow0 $(arrayname(A))" for A in NONDIAG_SQUARE_MATRICES
+        testJ0(pow0, A)
+        testH0(pow0, A)
+    end
+    @testset "pow0 $(arrayname(A))" for A in DIAG_SQUARE_MATRICES
+        # TODO: these should be zero and are currently too conservative
+        @test_broken all(iszero, Jsum(pow0, A))
+        @test_broken all(iszero, Hsum(pow0, A))
```

Comment:
To return empty Jacobians and Hessians on `pow(A, 0)`, we will need some additional methods on `Diagonal` matrices of tracers.

Reply:
Fixed in [bb16995](https://github.com/adrhill/SparseConnectivityTracer.jl/pull/165/commits/bb169953622662c8fdb9550c72c41f09e6fe3594)